src/kernels/MIOpenSigmoidFocalLoss.cpp

```diff
@@ -0,0 +1,329 @@
+ * Copyright (c) 2024 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+
+#ifndef MIOPEN_DONT_USE_HIP_RUNTIME_HEADERS
+#include <hip/hip_fp16.h>
+#include <hip/hip_runtime.h>
+#endif
+
+#include "float_types.h"
+#include "tensor_view.hpp"
+
+#ifndef IN_OUT_TYPE
+#define IN_OUT_TYPE float
+#endif
+
+#ifndef CVT_ACCUM2FLOAT
+#define CVT_ACCUM2FLOAT(x) (float_to_bfloat16(x))
+#endif
+
+#ifndef CVT_FLOAT2ACCUM
+#define CVT_FLOAT2ACCUM(x) (bfloat16_to_float(x))
+#endif
+
+template <typename TIO>
+__device__ void sigmoidFocalLossFwd(const TIO* input,
+                                    TIO* target,
+                                    TIO* workspace,
+                                    float alpha,
+                                    float gamma,
+                                    float divisor,
+                                    tensor_view_t<5> input_tv,
+                                    tensor_view_t<5> target_tv)
+{
+    size_t gid = threadIdx.x + blockIdx.x * blockDim.x;
+
+    tensor_layout_t<5> idx(input_tv, gid);
+    if(idx.layout[0] >= input_tv.size[0])
+        return;
+
+    FLOAT_ACCUM i = CVT_FLOAT2ACCUM(input[input_tv.get_tensor_view_idx(idx)]);
+    FLOAT_ACCUM t = CVT_FLOAT2ACCUM(target[target_tv.get_tensor_view_idx(idx)]);
+
+    FLOAT_ACCUM p      = 1 / (1 + exp(-i));
+    FLOAT_ACCUM ceLoss = -(t * log(p) + (1 - t) * log(1 - p));
+    FLOAT_ACCUM pT     = p * t + (1 - p) * (1 - t);
+    FLOAT_ACCUM loss   = ceLoss * pow(1 - pT, gamma);
+
+    if(alpha >= 0)
+    {
+        FLOAT_ACCUM alpha_t = alpha * t + (1 - alpha) * (1 - t);
+        loss                = alpha_t * loss;
+    }
+
+    workspace[gid] = CVT_ACCUM2FLOAT(loss / divisor);
+}
+
+extern "C" __global__ void SigmoidFocalLossFwd(const IN_OUT_TYPE* input,
+                                               IN_OUT_TYPE* target,
```

Comment:
Tensors are of the same dtype.
A dtype check of the tensor is required.
Please let me know if I can't find it. (I only found checkSameLength.)

Reply:
It's my bad. I have added the check code in problem description constructor